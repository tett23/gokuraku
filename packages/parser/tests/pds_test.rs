extern crate parser;

use insta::{assert_debug_snapshot, glob};
use std::fs;

#[test]
fn pds_test() {
    glob!("../fixtures/pds", "*.pds", |path| {
        let input = fs::read_to_string(path).unwrap();
        let ast = parser::prose_down_script_parse(&input).unwrap();

        assert_debug_snapshot!(ast);
    });
}
