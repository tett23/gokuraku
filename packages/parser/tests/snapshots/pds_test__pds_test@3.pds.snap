---
source: packages/parser/tests/pds_test.rs
expression: ast
input_file: packages/parser/fixtures/pds/3.pds
---
Module {
    statements: [
        AssignDef(
            AssignDef {
                eta_envs: EtaEnvs(
                    [],
                ),
                ident: Ident(
                    "(|)",
                ),
                expr: TypeExpr {
                    constraints: [],
                    expr: Arrow(
                        Literal(
                            Ident(
                                ForallIdent(
                                    ForallIdent(
                                        "a",
                                    ),
                                ),
                            ),
                        ),
                        Arrow(
                            Arrow(
                                Literal(
                                    Ident(
                                        ForallIdent(
                                            ForallIdent(
                                                "a",
                                            ),
                                        ),
                                    ),
                                ),
                                Literal(
                                    Ident(
                                        ForallIdent(
                                            ForallIdent(
                                                "b",
                                            ),
                                        ),
                                    ),
                                ),
                            ),
                            Literal(
                                Ident(
                                    ForallIdent(
                                        ForallIdent(
                                            "b",
                                        ),
                                    ),
                                ),
                            ),
                        ),
                    ),
                },
            },
        ),
        Assign(
            Assign {
                ident: Ident(
                    "(|)",
                ),
                args: AssignArgs {
                    patterns: [
                        Bind(
                            Ident(
                                "f",
                            ),
                        ),
                        Bind(
                            Ident(
                                "g",
                            ),
                        ),
                    ],
                },
                expr: Ident(
                    Ident(
                        "g",
                    ),
                ),
                where_clause: Module {
                    statements: [],
                },
            },
        ),
        AssignDef(
            AssignDef {
                eta_envs: EtaEnvs(
                    [],
                ),
                ident: Ident(
                    "echo",
                ),
                expr: TypeExpr {
                    constraints: [
                        Constraint {
                            ident: ExistsIdent(
                                ExistsIdent(
                                    "Show",
                                ),
                            ),
                            args: [
                                ForallIdent(
                                    ForallIdent(
                                        "a",
                                    ),
                                ),
                            ],
                        },
                    ],
                    expr: Arrow(
                        Literal(
                            Ident(
                                ForallIdent(
                                    ForallIdent(
                                        "a",
                                    ),
                                ),
                            ),
                        ),
                        Literal(
                            Context(
                                ExistsIdent(
                                    ExistsIdent(
                                        "IO",
                                    ),
                                ),
                                Literal(
                                    Tuple(
                                        0,
                                        [],
                                    ),
                                ),
                            ),
                        ),
                    ),
                },
            },
        ),
        Assign(
            Assign {
                ident: Ident(
                    "echo",
                ),
                args: AssignArgs {
                    patterns: [
                        Bind(
                            Ident(
                                "a",
                            ),
                        ),
                    ],
                },
                expr: Ident(
                    Ident(
                        "write",
                    ),
                ),
                where_clause: Module {
                    statements: [],
                },
            },
        ),
        AssignDef(
            AssignDef {
                eta_envs: EtaEnvs(
                    [],
                ),
                ident: Ident(
                    "main",
                ),
                expr: TypeExpr {
                    constraints: [],
                    expr: Literal(
                        Context(
                            ExistsIdent(
                                ExistsIdent(
                                    "IO",
                                ),
                            ),
                            Literal(
                                Tuple(
                                    0,
                                    [],
                                ),
                            ),
                        ),
                    ),
                },
            },
        ),
        Assign(
            Assign {
                ident: Ident(
                    "main",
                ),
                args: AssignArgs {
                    patterns: [],
                },
                expr: Abstruction(
                    Abstruction {
                        left: Abstruction(
                            Abstruction {
                                left: Ident(
                                    Ident(
                                        "|",
                                    ),
                                ),
                                right: Literal(
                                    Text(
                                        "text",
                                    ),
                                ),
                            },
                        ),
                        right: Ident(
                            Ident(
                                "echo",
                            ),
                        ),
                    },
                ),
                where_clause: Module {
                    statements: [],
                },
            },
        ),
    ],
}
